opendevin/runtime/plugins/jupyter/setup.sh

```diff
@@ -24,7 +24,23 @@ pip install jupyterlab notebook jupyter_kernel_gateway
 sudo mkdir -p /opendevin/logs && sudo chmod 777 /opendevin/logs
 
 # Run background process to start jupyter kernel gateway
-export JUPYTER_GATEWAY_PORT=18888
+# write a bash function that finds a free port
+find_free_port() {
+  local start_port="${1:-20000}"
+  local end_port="${2:-65535}"
+
+  for port in $(seq $start_port $end_port); do
+    if ! ss -tuln | awk '{print $5}' | grep -q ":$port$"; then
+      echo $port
+      return
+    fi
+  done
+
+  echo "No free ports found in the range $start_port to $end_port" >&2
+  return 1
+}
+
+export JUPYTER_GATEWAY_PORT=$(find_free_port 20000 30000)
```

Comment:
What is the necessity for finding a free port in a fresh sandbox?

Reply:
If you are using host network and spin up multiple sandboxes, fixing the port to a static one could cause port from different sandboxes to conflict each other, so that jupyter in some container would not work. This is a major issue i found while running multi-processing evaluation.